compiler/src/dmd/expressionsem.d

```diff
@@ -5335,7 +5335,8 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 lowering = new DotIdExp(exp.loc, lowering, Id.object);
 
                 auto tbn = exp.type.nextOf();
-                while (tbn.ty == Tarray)
+                size_t i = nargs;
+                while (tbn.ty == Tarray && --i)
```

Comment:
AFAICT, this is only used for `unqualTbn` below, which in turn is only used as 2nd template param for the `_d_newarraymTX` template. That template param `T` doesn't seem really well-defined; according to an example, it's supposed to be the final element type. But it additionally seems to be used once only (possibly fixed in the meantime?), further unqualified, in https://github.com/dlang/dmd/blob/ec25e56a93e1e6fbe9f3de1ab661cd0d40a82568/druntime/src/core/internal/array/construction.d#L532

So I don't think that `T` param is required, the element type should be trivial to infer.

Reply:
I agree, it would be better to remove this parameter and infer it in the runtime.
Can the interface between compiler and runtime be changed in a stable release?

Reply:
I'd be totally okay with that, compiler and druntime are a single, intertwined bundle.

Reply:
Wrt. inferring the 'base-element' in the template, that might be a bit more complicated than anticipated though - not sure if we can just recurse until the element type is no slice anymore (I guess multidimensional arrays of slices are a thing). And the length of the `size_t[] dims` runtime parameter isn't known at compile-time - I think your change here does the right thing, N `.nextOf()` calls for `N` specified dimensions.

Maybe it's simpler to keep an explicit `T`, or switching the `dims` parameter to a static `size_t[N]` array, with an `N` integer template parameter, which we could use to infer `T`.

Reply:
I prefer making `N` a template parameter and passing `size_t[N]`.

`_d_newarraymTX` is marked `@trusted`, but it is possible to pass `dims` with more dimensions than allowed in the type. The resulting array can then contain pointers with invalid values. Making `N` a template parameter should prevent that.

Reply:
Okay, as you wish. This is a pretty horrible bug, so I've approved of this first version already. Only nit: I'd replace the `tbn.ty == Tarray &&` check with an assert - we only expect to see slices while descending into the nested elements.

Reply:
I have created a separate issue for this: https://issues.dlang.org/show_bug.cgi?id=24514